ksql-api/src/test/java/io/confluent/ksql/api/ApiTest.java

```diff
@@ -0,0 +1,885 @@
+import java.util.function.Supplier;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ApiTest {
+
+  private static final long WAIT_TIMEOUT = 10000;
+  private static final JsonArray DEFAULT_COLUMN_NAMES = new JsonArray().add("name").add("age")
+      .add("male");
+  private static final JsonArray DEFAULT_COLUMN_TYPES = new JsonArray().add("STRING").add("INT")
+      .add("BOOLEAN");
+  private static final List<JsonArray> DEFAULT_ROWS = generateRows();
+
+  private Vertx vertx;
+  private Server server;
+  private TestEndpoints testEndpoints;
+  private WebClient client;
+
+  @Before
+  public void setUp() throws Throwable {
+
+    vertx = Vertx.vertx();
+
+    HttpServerOptions httpServerOptions =
+        new HttpServerOptions()
+            .setHost("localhost")
+            .setPort(8089)
+            .setUseAlpn(true)
+            .setSsl(true)
+            .setPemKeyCertOptions(
+                new PemKeyCertOptions().setKeyPath(findFilePath("test-server-key.pem"))
+                    .setCertPath(findFilePath("test-server-cert.pem"))
+            );
+
+    testEndpoints = new TestEndpoints(vertx);
+    JsonObject config = new JsonObject().put("verticle-instances", 4);
+    server = new Server(vertx, config, testEndpoints, httpServerOptions);
+    server.start();
+    client = createClient();
+    setDefaultRowGenerator();
+  }
+
+  @After
+  public void tearDown() {
+    if (client != null) {
+      client.close();
+    }
+    if (server != null) {
+      server.stop();
+    }
+  }
+
+  @Test
+  public void shouldExecutePullQuery() throws Exception {
+
+    JsonObject requestBody = new JsonObject().put("sql", "select * from foo").put("push", false);
+    JsonObject properties = new JsonObject().put("prop1", "val1").put("prop2", 23);
+    requestBody.put("properties", properties);
+
+    HttpResponse<Buffer> response = sendRequest("/query-stream", requestBody.toBuffer());
+
+    assertEquals(200, response.statusCode());
+    assertEquals("OK", response.statusMessage());
+    assertEquals("select * from foo", testEndpoints.getLastSql());
+    assertFalse(testEndpoints.getLastPush());
+    assertEquals(properties, testEndpoints.getLastProperties());
+
+    QueryResponse queryResponse = new QueryResponse(response.bodyAsString());
+    assertEquals(DEFAULT_COLUMN_NAMES, queryResponse.responseObject.getJsonArray("columnNames"));
+    assertEquals(DEFAULT_COLUMN_TYPES, queryResponse.responseObject.getJsonArray("columnTypes"));
+    assertEquals(DEFAULT_ROWS, queryResponse.rows);
+    assertEquals(0, server.getQueryIDs().size());
+    String queryID = queryResponse.responseObject.getString("queryID");
+    assertNotNull(queryID);
+    assertFalse(server.getQueryIDs().contains(queryID));
+    Integer rowCount = queryResponse.responseObject.getInteger("rowCount");
+    assertNotNull(rowCount);
+    assertEquals(DEFAULT_ROWS.size(), rowCount.intValue());
+  }
+
```

Comment:
Tbh, I find the given, when, then thing quite forced and simplistic. I can add it for some of the simpler tests.
But other tests won't fit so easily into this pattern. E.g. in some test I do things like start a bunch of push queries then do some asserts on the state, then I close the queries and do another bunch of asserts on the state. In this case there are multiple "whens" in the test .

Reply:
I've found the Given/When/Then pattern a forcing function to make tests simple. There are admittedly some tests that are way more difficult to write when following that pattern, but those should be the exception not the rule IMO.

Reply:
Yep, that's its purpose. If you've got a test that doesn't fit, it generally means you're trying to test too much in one test and it would be better split into multiple.